storage/innobase/log/log0online.cc

```diff
@@ -1370,13 +1372,16 @@ is tracked.
 bool log_online_follow_redo_log(void) {
   ut_ad(!srv_read_only_mode);
 
-  const auto last_checkpoint_lsn = log_get_checkpoint_lsn(*log_sys);
+  const auto last_checkpoint_aligned_lsn = ut_uint64_align_up(
+      log_get_checkpoint_lsn(*log_sys), OS_FILE_LOG_BLOCK_SIZE);
   bool result = true;
```

Comment:
So this is the core change. We are covering bytes after checkpoint_lsn.

But we are using the entire block which may or may not have redo records.

Is it possible to determine the LSN at which the current redo record ends? (I think recovery also tries to find the end of redo record from checkpoint LSN).

Reply:
Btw, Sorry for the late response..  :(

Reply:
To my understanding we can not predict where log data will end without parsing that data. What I can see in recovery we still read some number of full blocks from log from `start_lsn` to `start_lsn + 4 * UNIV_PAGE_SIZE` and then stop parsing that data once we find partially filled in block.

Reply:
yes and no. There is no prediction as such. Like you said, we have to parse redo records and find the start of first record after the checkpoint_lsn.

Look for "parse_start_lsn" in recovery code. So essentially we need parse_start_lsn which will include bytes after the checkpoint_lsn and before the start of next record.